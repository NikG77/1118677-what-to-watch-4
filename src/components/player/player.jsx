import React from "react";
import PropTypes from "prop-types";
import {formateDuration} from "../../utils/utils.js";
import history from "../../history.js";
// import Loader from "../loader/loader.jsx";
import {AppRoute} from "../../const.js";
import {filmType} from "../../types/types";

const Player = (props) => {
  const {id, isPlay, duration, progress, onPlayClick, onFullScreenClick, forwardedRef, setDuration} = props;

  return (
    <div className="player">

      <video
        ref={forwardedRef}
        onLoadedMetadata={setDuration}
        className="player__video"
      />

      <button type="button" className="player__exit"
        onClick={() => history.push(`${AppRoute.FILM}/${id}`)}>Exit</button>

      <div className="player__controls">
        <div className="player__controls-row">
          <div className="player__time">
            <progress className="player__progress" value={progress} max="100"></progress>
            <div className="player__toggler" style={{left: `${progress}%`}}>Toggler</div>
          </div>
          <div className="player__time-value">{formateDuration(duration * (100 - progress) / 100)}</div>
        </div>

        <div className="player__controls-row">
          <button onClick={onPlayClick} type="button" className="player__play">

            {isPlay || <React.Fragment><svg viewBox="0 0 19 19" width="19" height="19">
              <use xlinkHref="#play-s"></use>
            </svg>
            <span>Play</span></React.Fragment>}

            {isPlay && <React.Fragment>
              <svg viewBox="0 0 14 21" width="14" height="21">
                <use xlinkHref="#pause"></use>
              </svg>
              <span>Pause</span></React.Fragment>}

          </button>
          <div className="player__name">Transpotting</div>

          <button onClick={onFullScreenClick} type="button" className="player__full-screen">
            <svg viewBox="0 0 27 27" width="27" height="27">
              <use xlinkHref="#full-screen"></use>
            </svg>
            <span>Full screen</span>
          </button>
        </div>
      </div>

    </div>

  );

};


Player.propTypes = {
  isPlay: PropTypes.bool.isRequired,
  duration: PropTypes.number.isRequired,
  progress: PropTypes.number.isRequired,
  onPlayClick: PropTypes.func.isRequired,
  onFullScreenClick: PropTypes.func.isRequired,
  setDuration: PropTypes.func.isRequired,
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any})
  ]),
  id: PropTypes.oneOfType([
    () => null,
    PropTypes.number.isRequired,
  ]),
  film: PropTypes.oneOfType([
    filmType.isRequired,
    PropTypes.oneOf([null]).isRequired,
  ]),
};

const PlayerWrap = React.forwardRef((props, ref) => <Player {...props} forwardedRef={ref} />);

export {Player};
export default PlayerWrap;
