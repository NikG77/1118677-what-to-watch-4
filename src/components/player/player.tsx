import * as React from "react";

import {AppRoute} from "../../const";
import {formateDuration} from "../../utils/utils";
import history from "../../history";

interface Props {
  duration: number;
  id: number;
  isPlay: boolean;
  forwardedRef: any;
  onFullScreenClick: () => void;
  onPlayClick: () => void;
  progress: number;
  setDuration: () => void;
}

const Player: React.FunctionComponent<Props> = (props: Props) => {
  const {
    duration,
    id,
    isPlay,
    forwardedRef,
    onFullScreenClick,
    onPlayClick,
    progress,
    setDuration
  } = props;

  return (
    <div className="player">

      <video
        ref={forwardedRef}
        onLoadedMetadata={setDuration}
        className="player__video"
      />

      <button type="button" className="player__exit"
        onClick={() => history.push(`${AppRoute.FILM}/${id}`)}>Exit</button>

      <div className="player__controls">
        <div className="player__controls-row">
          <div className="player__time">
            <progress className="player__progress" value={progress} max="100"></progress>
            <div className="player__toggler" style={{left: `${progress}%`}}>Toggler</div>
          </div>
          <div className="player__time-value">{formateDuration(duration * (100 - progress) / 100)}</div>
        </div>

        <div className="player__controls-row">
          <button onClick={onPlayClick} type="button" className="player__play">

            {isPlay || <React.Fragment><svg viewBox="0 0 19 19" width="19" height="19">
              <use xlinkHref="#play-s"></use>
            </svg>
            <span>Play</span></React.Fragment>}

            {isPlay && <React.Fragment>
              <svg viewBox="0 0 14 21" width="14" height="21">
                <use xlinkHref="#pause"></use>
              </svg>
              <span>Pause</span></React.Fragment>}

          </button>
          <div className="player__name">Transpotting</div>

          <button onClick={onFullScreenClick} type="button" className="player__full-screen">
            <svg viewBox="0 0 27 27" width="27" height="27">
              <use xlinkHref="#full-screen"></use>
            </svg>
            <span>Full screen</span>
          </button>
        </div>
      </div>

    </div>

  );

};



const PlayerWrap = React.forwardRef<React.FunctionComponent, Props>((props, ref) => <Player {...props} forwardedRef={ref} />);

export {Player};
export default PlayerWrap;
