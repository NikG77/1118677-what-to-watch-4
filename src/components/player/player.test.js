import React from "react";
import renderer from "react-test-renderer";
import {Player} from "./player";

it(`Should Player render correctly`, () => {
  const tree = renderer
    .create(<Player
      onExitPlayButtonClick={() => {}}
      isPlay={true}
      duration={30}
      progress={10}
      onPlayClick={() => {}}
      onFullScreenClick={() => {}}
      setDuration={() => {}}
    />,
    {
      createNodeMock: () => {
        return {};
      }
    }).toJSON();

  expect(tree).toMatchSnapshot();
});
