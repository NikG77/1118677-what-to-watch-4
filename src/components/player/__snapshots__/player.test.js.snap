// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Player render correctly 1`] = `
<p
  style={
    Object {
      "fontSize": "40px",
      "textAlign": "center",
    }
  }
>
  Loading...
</p>
`;
